Add a few new features

rneswold opened the pull request:
- Finished the Plot Configuration API
- The JWT is forwarded to the gRPC requests. This is a work-in-progress and shouldn't be considered stable yet.
- Readings devices now supports 4 device values: `DevScalar`, `DevScalarArray`, `DevText`, and `DevTextArray`

rneswold committed:
:bug: fix race condition

It seemed that we could never revoke the credentials. This code
revoked them in a microtask. However, futures don't make progress
until the main thread yeilds. Instead of yielding, we immediately
call `setState()` which `null`s the credentials before we revoke
them.

This version saves a reference so the microtask can actually
revoke the credentials.

rneswold committed:
:arrow_up: specify latest Flutter SDK

This uses the latest linter and fixes some new warnings.

rneswold committed:
:fire: we don't actually use the `ferry_flutter` package

rneswold committed:
:arrow_up: use latest packages

rneswold committed:
:sparkles: finish user plot config changes

This adds two, new queries to the schema.

- `retrieveLastUserConfiguration()` gets the last config the application
  saved for this user.
- `saveUserConfiguration(cfg)` takes the given config and saves it as
  the user's current, start-up config.

We also take this opportunity to rebuild the GraphQL interface code
using the latest tools that were brought in from the previous commit.

rneswold committed:
:sparkles: forward JWT to GraphQL service

rneswold committed:
:sparkles: support more device values

When reading devices, the API now returns `DevScalar`, `DevScalarArray`,
`DevText`, and `DevTextArray`.

rneswold commented:
This will be 0.2.9 and will have breaking changes since `Reading`s return more than just a `double`.

rneswold committed:
:recycle: add helper extensions for DevValues

These extensions take a DevValue and convert them to the compatible,
primitive type.

rneswold committed:
:recycle: use `null` instead of exceptions

I immediately reached for exception handling when implementing these
extensions. But that just makes the calling code more complicated. In
this commit, we return `null` instead of throwing an exception.

I applied this API change to the `monitorDigitalStatus` query and it
cleaned up reading the value.

beauremus commented on lib/src/service/acsys_service.dart:
```diff
@@ -371,7 +371,7 @@ final class ChannelSettingSnapshot {
 // Only used by the plot ID class to generate IDs for testing.
 
-int _genPlotId = 1_000_000;
+int _genPlotId = 1000000;
```
Why remove the clarity of the separators?

rneswold replied:
Because VSCode was complaining about it -- like it was using an older Dart SDK. I couldn't figure out why it was complaining, but I like the separators.

rneswold replied:
Hmmmm. Seems to be working now. 🤨

I'll fix it in a new commit.

beauremus commented on lib/src/service/acsys_service.dart:
```diff
@@ -501,12 +505,16 @@ final class ACSysService implements ACSysServiceAPI {
+            link: HttpLink("https://acsys-proxy.fnal.gov:8001/acsys",
+                defaultHeaders: _buildAuthHeader(jwt)),
             cache: Cache()),
         _s = Client(
```
I think the WebSocket should need the JWT as well, right?
If that's the case, would it be good to add the AuthHeaders in the `Client` implementation since it applies everywhere?

rneswold replied:
Good catch! Yes, we definitely want them there, too. But the WebSocket API is a different beast and I didn't know how to pass the credentials using it. Since settings use the "mutation" part of GraphQL, I felt I could proceed.

rneswold replied:
The GraphQL framework hides the upgrade to a WebSocket, so I have to dig in the docs to see if I can gain access to the http headers.

beauremus commented on lib/src/service/acsys_service.dart:
```diff
@@ -1119,16 +1147,27 @@ class ACSysProvider extends InheritedWidget {
+  }
+}
+
+final class _ACSysProviderIW extends InheritedWidget {
```
I think this is to elevate state, but its not clear.

rneswold replied:
I'll add comments. Thanks.

rneswold committed:
:art: put separators in large numbers

rneswold committed:
:art: clean-up for `_ACSysProviderIW`

- Added comments explaining why we need it
- Removed a public property that no client could use (but the class is
  private.)

beauremus reviewed (approved):
LGTM! :shipit:
